### PR TITLE
LLVM Submodule Update

### PR DESCRIPTION
**Pull request by uenoku**
Close https://github.com/llvm/circt/issues/2195

Updates:
- [x]  Replace `Identifier` with `StringAttr` (https://reviews.llvm.org/D113536)
- [x]  Use` mlir::NamedAttribute` (https://reviews.llvm.org/D113956)
- [x]  Use` macher::m_Any` (https://reviews.llvm.org/D113905)
- [x]   Update printer/parser of types/attributes (https://reviews.llvm.org/D113637) to use AsmPrinter/Parser instead of DialectAsmParser/Printer

### CODE REVIEW

**Comment by seldridge** [comment]
Thanks for taking initiative with this, @uenoku!

**Comment by uenoku** [comment]
It seems assembly formats for types and attributes are supported(https://reviews.llvm.org/D111594) so there are related changes (i.e. we don't have to print  mnemonic anymore)

**Comment by uenoku** [comment]
I'm down to one test left, but I would stop it until after the vacation 😃

**Comment by uenoku** [comment]
It seems the problem is that outputs of inst/entity/proc outlive the conversion.

**Comment by maerhart** [comment]
Thanks for updating LLVM, @uenoku ! And also for providing the LLVM commit leading to the issue!

**Comment by uenoku** [comment]
Thank you for fixing this, @maerhart!!

**Comment by uenoku** [comment]
clang-tidy fails strangely even though the build passes. I guess it would be some false-positive so I think it's fine with it.

**Comment by lattner** [comment]
Yeah, clang-tidy does have false positives.

**Comment by seldridge** [comment]
I'd like to get this in, but am lost as to what `clang-tidy` is actually complaining about. 😕 Does anybody know what is the right way to proceed with this?

**Comment by lattner** [comment]
Clang tidy is complaining about build errors.  I'd just force land this.  Thank you for working on this!

**Comment by teqdruid** [comment]
I think tidy is automatically removing `using mlir::AsmParser`, then detecting that it now doesn't exist in all the files which don't use it. I would suggest [disabling all `clang-tidy` checks](https://clang.llvm.org/extra/clang-tidy/#suppressing-undesired-diagnostics) in LLVM.h. E.g.:

```
// NOLINTBEGIN(misc-unused-using-decls)
... all usings
// NOLINTEND(misc-unused-using-decls)
```

This _might_ work.

I'm not sure why this is being triggered by this PR.

**Comment by uenoku** [comment]
> I think tidy is automatically removing `using mlir::AsmParser`, then detecting that it now doesn't exist in all the files which use it. I would suggest [disabling all `clang-tidy` checks](https://clang.llvm.org/extra/clang-tidy/#suppressing-undesired-diagnostics) in LLVM.h. E.g.:
> 
> ```
> // NOLINTBEGIN(misc-unused-using-decls)
> ... all usings
> // NOLINTEND(misc-unused-using-decls)
> ```
> 

It worked! Thanks @teqdruid!

**Comment by uenoku** [comment]
Oh, it seems `NOLINTBEGIN/END` is supported from clang-tidy-13 😅 The current version in docker image is 12 though.

**Comment by uenoku** [comment]
Integration tests are failing. Will take a look at it.

**Comment by uenoku** [comment]
It seems some mlir python update got to make verification work at the creation of operations (previously not).
For example, `/usr/bin/python3.9 ./circt/integration_test/Bindings/Python/dialects/comb.py ` is currently generating
```
    %c1_i32 = hw.constant 1 : i32
    %true = hw.constant true
    %0 = comb.extract %c1_i32 from 14 : (i32) -> i32
    %1 = comb.extract %c1_i32 from 14 : (i32) -> i32
```
but this is wrong.

**Comment by uenoku** [comment]
> Oh, it seems NOLINTBEGIN/END is supported from clang-tidy-13 (https://reviews.llvm.org/D108560) 😅 We need to update the docker image because the current version in docker image 12. I updated to use clang-tidy-13. 

Sorry I mistook the version, `NOLINTBEGIN/END` is not included in clang-tidy-13;) So I'd like to just ignore clang-tidy failure. (updating llvm in CI is good so I'd like to leave it)

**Comment by teqdruid** [comment]
> So I'd like to just ignore clang-tidy failure.

My only concern is other PRs hitting this, but given that you changed those two lines I don't think that'll happen. So I'm good with that.

**Comment by uenoku** [comment]
One option is to add `NOLINT(misc-unused-using-decls)` to all usings like,
```
using mlir::ArrayAttr;  // NOLINT(misc-unused-using-decls)
using mlir::AsmParser; // NOLINT(misc-unused-using-decls)
using mlir::AsmPrinter; // NOLINT(misc-unused-using-decls)
using mlir::Attribute; // NOLINT(misc-unused-using-decls)
using mlir::Block; // NOLINT(misc-unused-using-decls)
using mlir::BlockAndValueMapping; // NOLINT(misc-unused-using-decls)
....
```

**Comment by seldridge** [comment]
I'm fine with inline `NOLINT`.

**Comment by uenoku** [comment]
> I'm fine with inline NOLINT.

Thanks, I changed so

File: lib/Dialect/FIRRTL/FIRRTLTypes.cpp [hunk 1]
```diff
@@ -965,14 +962,14 @@ std::pair<unsigned, bool> FVectorType::rootChildFieldID(unsigned fieldID,
 // CMemory Type
 //===----------------------------------------------------------------------===//
 
-void CMemoryType::print(mlir::DialectAsmPrinter &printer) const {
-  printer << "cmemory<";
+void CMemoryType::print(AsmPrinter &printer) const {
+  printer << "<";
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLTypes.cpp:966]: The "cmemory" string is now automatically printed, yes?

File: include/circt/Support/LLVM.h [hunk 2]
```diff
@@ -167,97 +169,102 @@ struct OpRewritePattern;
...
-using mlir::Value;
-using mlir::ValueRange;
-using mlir::VectorType;
-using mlir::WalkResult;
+// clang-tidy removes following using directives incorrectly. So force
+// clang-tidy to ignore them.
+// TODO: It is better to use `NOLINTBEGIN/END` comments to disable clang-tidy
+// than adding `NOLINT` to every line. `NOLINTBEGIN/END` will supported from
+// clang-tidy-14.
+using mlir::ArrayAttr;                 // NOLINT(misc-unused-using-decls)
```
Comment (seldridge) [inline include/circt/Support/LLVM.h:184]: This is unfortunately ugly, but I don't know how else to handle it and us fighting with clang-tidy on code that _clearly compiles and is fine_ shouldn't hold up the PR.  We can try to figure out how to remove this in the future.

**Review by seldridge — APPROVED** [review]
Awesome. Thanks for fighting with this, @uenoku!

File: lib/Dialect/FIRRTL/FIRRTLTypes.cpp (hunk 1 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/FIRRTLTypes.cpp:966]: Yes, mnemonic of type is now automatically printed.

File: include/circt/Support/LLVM.h (hunk 2 above)
Comment (uenoku) [reply include/circt/Support/LLVM.h:184]: Yes, agreed. It's definitely a bug of `misc-unused-using-decls`.

**Comment by teqdruid** [comment]
I suspect it's since clang-tidy doesn't expect to see 'using's in a header.
(Since it's not generally considered good practice.)

**Pull request merged**